File: tags/gifteng-front/0.9/application/libraries/usermanagement_service.php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

/**
 * User management service
 *
 * @author gyuszi
 */
class Usermanagement_service {
    
    var $userService;
    var $token;
    
    public function __construct() {
        log_message(DEBUG, "Initializing Usermanagement_service");
    }
    
    //*****************************
    //* user verification related *
    //*****************************
    
    /**
     * 
     * @param string $code
     * @throws Exception
     */
    public function verifyUser($code) {
        try {
            $userService = $this->getService();
            $userService->verifyUser(array("code" => $code));
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @throws Exception
     */
    public function resendVerification() {
        try {
            $userService = $this->getService();
            $userService->resendVerification();
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //**********************
    //* categories related *
    //**********************
    
    /**
     * 
     * @return array of BusinessCategory_model
     * @throws Exception
     */
    public function getAllBusinessCategories() {
        try {
            $userService = $this->getService();
            $result = $userService->getAllBusinessCategories();
            
            $categories = array();
            if ( hasField($result, 'category') && $result->category ) {
                $categories = BusinessCategory_model::convertBusinessCategories($result->category);
            }
            return $categories;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Getting all business categories failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //************************************
    //* user crud (create/update/delete) *
    //************************************
    
    /**
     * Registers business user.
     * 
     * @param User_model $user
     * @param string $password
     * @return long the created user ID
     * @throws Exception
     */
    public function registerBusinessUser($user, $password) {
        try {
            $userService = $this->getService();
            $result = $userService->registerBusinessUser(array(
                "user" => $user,
                "password" => $password
            ));
            return $result->userId;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Invokes the user registration on WS.
     * 
     * @param User_model $user
     * @param string $password
     * @param string $code
     * @return long
     * @throws Exception in case of WS error
     */
    public function registerUser($user, $password, $code) {
        try {
            $userService = $this->getService();
            $result = $userService->registerUser(array(
                "user" => $user,
                "password" => $password,
                "invitationCode" => $code
            ));
            return $result->userId;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw $ex;
        }
    }
    
    /**
     * 
     * @param User_model $user
     * @return boolean true if all required information is gathered.
     * @throws Exception
     */
    public function updateUser($user) {
        try {
            $userService = $this->getService();
            $result = $userService->updateUser(array("user" => $user));
            return $result->complete;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Put the currently logged user into deactivated state.
     * 
     * @throws Exception
     */
    public function deactivateUser() {
        try {
            $userService = $this->getService();
            $userService->deactivateUser();
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }


    //*****************
    //* user statistics *
    //*****************
    
    public function getStatistics($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->getStatistics(array("userId" => $userId));
            $statistics = UserStatistics_model::convertUserStatistics($result->statistics);
            return $statistics;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'User statistics (id: '.$userId.') request failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //***************
    //* user search *
    //***************
    
    /**
     * Returns a list of users having the highest score.
     * 
     * @param type $numberUsers
     * @return type
     * @throws Exception
     */
    public function getTopUsers($numberUsers) {
        try {
            $userService = $this->getService();
            $result = $userService->getTopUsers(array("numberUsers" => $numberUsers));
            
            $users = array();
            if ( hasField($result, 'users') && $result->users ) {
                $users = User_model::convertUsers($result->users);
            }
            return $users;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Request user by its user name;
     * 
     * @param string $name
     * @return User_model
     * @throws Exception
     */
    public function getUserByName($name) {
        try {
            $userService = $this->getService();
            $result = $userService->getUserByName(array("name" => $name));
            $user = $result->user;
            return User_model::convertUser($user);
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $userId
     * @return User_model
     * @throws Exception
     */
    public function getUserById($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->getUserById(array("userId" => $userId));
            $user = $result->user;
            return User_model::convertUser($user);
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //***************
    //* user follow *
    //***************
    
    /**
     * Follow the given user.
     * 
     * @param long $userId
     * @return UserStatistics_model
     * @throws Exception if user not found or WS error
     */
    public function follow($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->follow(array("userId" => $userId));
            $statistics = UserStatistics_model::convertUserStatistics($result->statistics);
            return $statistics;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Unfollow the given user.
     * 
     * @param long $userId
     * @return UserStatistics_model
     * @throws Exception if user not found or WS error
     */
    public function unfollow($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->unfollow(array("userId" => $userId));
            $statistics = UserStatistics_model::convertUserStatistics($result->statistics);
            return $statistics;
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Requests the followers list - the ones that logged/current user is following.
     * 
     * @param long $userId
     * @return list of User_model
     * @throws Exception
     */
    public function getFollowers($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->getFollowers(array("userId" => $userId));
            
            $users = array();
            if ( hasField($result, 'follower') && $result->follower ) {
                $users = User_model::convertUsers($result->follower);
            }
            return $users;
        } catch ( Exception $ex ) {
            log_message(ERROR, "User followers request failed! " . $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * Returns the followings list.
     * 
     * @param long $userId
     * @return list of User_model
     * @throws Exception
     */
    public function getFollowings($userId) {
        try {
            $userService = $this->getService();
            $result = $userService->getFollowings(array("userId" => $userId));
            
            $users = array();
            if ( hasField($result, 'following') && $result->following ) {
                $users = User_model::convertUsers($result->following);
            }
            return $users;
        } catch ( Exception $ex ) {
            log_message(ERROR, "User followings request failed! " . $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //*****************
    //* user settings *
    //*****************
    
    /**
     * 
     * @return UserSetting_model
     * @throws Exception
     */
    public function getUserSetting() {
        try {
            $userService = $this->getService();
            $result = $userService->getUserSetting();
            
            $userSetting = UserSetting_model::convertUserSetting($result->setting);
            return $userSetting;
        } catch ( Exception $ex ) {
            log_message(ERROR, "User setting request failed! " . $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param UserSetting_model $userSetting
     * @throws Exception
     */
    public function saveUserSetting($userSetting) {
        try {
            $userService = $this->getService();
            $userService->saveUserSetting(array("setting" => $userSetting));
        } catch ( Exception $ex ) {
            log_message(ERROR, "Saving user setting failed! " . $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //******************
    //* social network *
    //******************
    
    /**
     * 
     * @return array of string
     * @throws Exception
     */
    public function getConnectedSocialNetworks() {
        try {
            $userService = $this->getService();
            $result = $userService->getConnectedSocialNetworks();
            
            $networks = array();
            if ( hasField($result, 'network') && $result->network ) {
                if ( is_array($result->network) && count($result->network) > 0 ) {
                    foreach ( $result->network as $network ) {
                        array_push($networks, $network);
                    }
                } elseif ( !is_empty($result->network) ) {
                    $network = $result->network;
                    array_push($networks, $network);
                }
            }
            return $networks;
        } catch ( Exception $ex ) {
            log_message(ERROR, "Getting connected social network list failed! " . $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    
    
    /**
     * Gets the user from the server by the given email and stores into session.
     * 
     * @param string $email the user email
     * @param string $token
     * @throws Exception in case of WS invocation error
     */
    public function storeUserByEmail($email, $token) {
        $user = $this->getUserByEmail($email, $token);
        $this->storeUser($user);
    }
    
    /**
     * Stores the given user into session under 'user' key.
     * 
     * @param User_model $user
     */
    public function storeUser($user) {
        storeIntoSession('user', $user);
    }
    
    /**
     * Loads the user from the session that was stored previously.
     * 
     * @return User_model
     */
    public function loadUser() {
        return loadFromSession('user');
    }
    
    /**
     * Refresh the user in the session by requesting server.
     * 
     * @return User_model
     */
    public function refreshUser() {
        $user = $this->loadUser();
        $token = loadToken();
        $this->storeUserByEmail($user->email, $token);
        return $this->loadUser();
    }
    
    /**
     * Refresh the currently logged in user statistics.
     * 
     * @return User_model
     */
    public function refreshStatistics() {
        $user = $this->loadUser();
        
        try {
            $statistics = $this->getStatistics($user->id);
            $user->statistics = $statistics;
            $this->storeUser($user);
        } catch ( Exception $ex ) {
        }
        
        return $user;
    }
    
    /* internal functions */
    
    private function getUserByEmail($email, $token) {
        try {
            $userService = $this->getService($token);
            $result = $userService->getUserByEmail(array("email" => $email));
            $user = $result->user;
            return User_model::convertUser($user);
        } catch ( Exception $ex ) {
            log_message(ERROR, $ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    private function getService($token = null) {
        if ( $this->userService == null || $this->token != $token ) {
            if ( $token == null ) {
                $token = loadToken();
            }
            
            $this->token = $token;
            $this->userService = new SoapClient(USER_SERVICE_WSDL, getSoapOptions($this->token));
        }
        return $this->userService;
    }
}

File: trunk/gifteng-webapp/src/main/java/com/venefica/service/UserManagementService.java
package com.venefica.service;

import com.venefica.service.dto.BusinessCategoryDto;
import com.venefica.service.dto.UserDto;
import com.venefica.service.dto.UserSettingDto;
import com.venefica.service.dto.UserStatisticsDto;
import com.venefica.service.fault.GeneralException;
import com.venefica.service.fault.InvalidInvitationException;
import com.venefica.service.fault.InvitationNotFoundException;
import com.venefica.service.fault.UserAlreadyExistsException;
import com.venefica.service.fault.UserNotFoundException;
import java.util.List;
import javax.jws.WebMethod;
import javax.jws.WebParam;
import javax.jws.WebResult;
import javax.jws.WebService;
import javax.jws.soap.SOAPBinding;
import javax.jws.soap.SOAPBinding.ParameterStyle;
import javax.validation.constraints.NotNull;

/**
 * Registers new users and updates information about them.
 *
 * @author Sviatoslav Grebenchukov
 */
@WebService(name = "UserManagement", targetNamespace = Namespace.SERVICE)
@SOAPBinding(parameterStyle = ParameterStyle.WRAPPED)
public interface UserManagementService {
    
    //*****************************
    //* user verification related *
    //*****************************
    
    @WebMethod(operationName = "VerifyUser")
    void verifyUser(@WebParam(name = "code") @NotNull String code) throws UserNotFoundException, GeneralException;
    
    @WebMethod(operationName = "ResendVerification")
    void resendVerification() throws UserNotFoundException, GeneralException;
    
    
    
    //**********************
    //* categories related *
    //**********************
    
    @WebMethod(operationName = "GetAllBusinessCategories")
    @WebResult(name = "category")
    List<BusinessCategoryDto> getAllBusinessCategories();
    
    
    
    //************************************
    //* user crud (create/update/delete) *
    //************************************
    
    /**
     * Registers a new business user.
     * 
     * @param userDto the user to be created
     * @param password
     * @return
     * @throws UserAlreadyExistsException when a user with the same email exists
     * @throws GeneralException when the provided category does not exists
     */
    @WebMethod(operationName = "RegisterBusinessUser")
    @WebResult(name = "userId")
    public Long registerBusinessUser(
            @WebParam(name = "user") @NotNull UserDto userDto,
            @WebParam(name = "password") @NotNull String password) throws UserAlreadyExistsException, GeneralException;
    
    /**
     * Registers new local user.
     *
     * @param userDto the user to register
     * @param password the password (may be a hash of the password in the future)
     * @param invitationCode the invitation code
     * @param referrerId the referrer user id
     * @param shareId the share id
     * @throws UserAlreadyExistsException is thrown when a user with the same
     * name and email already exists
     * @throws InvitationNotFoundException thrown when the provided invitation code
     * could not be found
     * @throws GeneralException if the user verification code could not be created
     */
    @WebMethod(operationName = "RegisterUser")
    @WebResult(name = "userId")
    public Long registerUser(
            @WebParam(name = "user") @NotNull UserDto userDto,
            @WebParam(name = "password") @NotNull String password,
            @WebParam(name = "invitationCode")  @NotNull String invitationCode,
            @WebParam(name = "referrerId")  Long referrerId,
            @WebParam(name = "shareId")  Long shareId)
            throws UserAlreadyExistsException, InvitationNotFoundException, InvalidInvitationException, GeneralException;
    
    /**
     * Updates user information.
     *
     * @param userDto updated user object
     * @return true if all required information is gathered.
     */
    @WebMethod(operationName = "UpdateUser")
    @WebResult(name = "complete")
    public boolean updateUser(@WebParam(name = "user") @NotNull UserDto userDto)
            throws UserNotFoundException, UserAlreadyExistsException;
    
    /**
     * Returns true if all required information is gathered for the current
     * user.
     */
    @WebMethod(operationName = "IsUserComplete")
    @WebResult(name = "complete")
    public boolean isUserComplete() throws UserNotFoundException;
    
    /**
     * Mark the currently logged user as deactivated (deleted). User and it's
     * gifts will not be show anymore.
     */
    @WebMethod(operationName = "DeactivateUser")
    public void deactivateUser() throws UserNotFoundException;

    
    
    //*******************
    //* user statistics *
    //*******************
    
    @WebMethod(operationName = "GetStatistics")
    @WebResult(name = "statistics")
    public UserStatisticsDto getStatistics(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;
    
    
    
    //***************
    //* user search *
    //***************
    
    /**
     * Retrieves a list of users having the highest score.
     * 
     * @param numberUsers
     * @return 
     */
    @WebMethod(operationName = "GetTopUsers")
    @WebResult(name = "users")
    public List<UserDto> getTopUsers(@WebParam(name = "numberUsers") int numberUsers) throws UserNotFoundException;
    
    /**
     * Retrieves information about the current (logged) user.
     *
     * @return user object
     * @throws UserNotFoundException is thrown when a user with the specified
     * token (id) not found
     */
    @WebMethod(operationName = "GetUser")
    @WebResult(name = "user")
    public UserDto getUser() throws UserNotFoundException;

    /**
     * Retrieves information about the user by his name.
     *
     * @param name name of the user
     * @return user object
     * @throws UserNotFoundException is thrown when a user with the specified
     * name not found
     */
    @WebMethod(operationName = "GetUserByName")
    @WebResult(name = "user")
    public UserDto getUserByName(@WebParam(name = "name") @NotNull String name) throws UserNotFoundException;
    
    /**
     * Retrieves information about the user by his email address.
     * 
     * @param email the email address of the user
     * @return user object
     * @throws UserNotFoundException when the user with the specified email
     * address not found
     */
    @WebMethod(operationName = "GetUserByEmail")
    @WebResult(name = "user")
    public UserDto getUserByEmail(@WebParam(name = "email") @NotNull String email) throws UserNotFoundException;
    
    /**
     * Retrieves information about the user by his phone number.
     * 
     * @param phone the phone number of the user
     * @return user object
     * @throws UserNotFoundException when the user with the specified phone
     * number not found
     */
    @WebMethod(operationName = "GetUserByPhone")
    @WebResult(name = "user")
    public UserDto getUserByPhone(@WebParam(name = "phone") @NotNull String phone) throws UserNotFoundException;
    
    /**
     * 
     * @param userId
     * @return
     * @throws UserNotFoundException 
     */
    @WebMethod(operationName = "GetUserById")
    @WebResult(name = "user")
    public UserDto getUserById(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;

    
    
    //***************
    //* user follow *
    //***************
    
    /**
     * Adds the given user into the actual ones followers list.
     * 
     * @param userId the user identifier
     * @return the current user statistics
     * @throws UserNotFoundException the given user could not be found
     */
    @WebMethod(operationName = "Follow")
    @WebResult(name = "statistics")
    public UserStatisticsDto follow(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;
    
    /**
     * Removes the given user from the followers list of the actual one.
     * 
     * @param userId the user identifier
     * @return the current user statistics
     * @throws UserNotFoundException the given user could not be found
     */
    @WebMethod(operationName = "Unfollow")
    @WebResult(name = "statistics")
    public UserStatisticsDto unfollow(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;
    
    /**
     * Returns the list of users that are following the given one.
     * 
     * @param userId
     * @return list of followers
     * @throws UserNotFoundException if the given user could not be found
     */
    @WebMethod(operationName = "GetFollowers")
    @WebResult(name = "follower")
    public List<UserDto> getFollowers(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;
    
    /**
     * Returns a list of users that the given one is following.
     * 
     * @param userId
     * @return list of followings
     * @throws UserNotFoundException if the given user could not be found
     */
    @WebMethod(operationName = "GetFollowings")
    @WebResult(name = "following")
    public List<UserDto> getFollowings(@WebParam(name = "userId") @NotNull Long userId) throws UserNotFoundException;
    
    
    
    //*****************
    //* user settings *
    //*****************
    
    /**
     * Returns the current user settings.
     * 
     * @return
     * @throws GeneralException if the user is not member type
     */
    @WebMethod(operationName = "GetUserSetting")
    @WebResult(name = "setting")
    public UserSettingDto getUserSetting() throws UserNotFoundException, GeneralException;
    
    /**
     * Saves/updates the given settings for the given and current user.
     * 
     * @param userSettingDto
     * @throws GeneralException if the current user is business type, or the current
     * user is not the same as the updating one
     */
    @WebMethod(operationName = "SaveUserSetting")
    public void saveUserSetting(@WebParam(name = "setting") @NotNull UserSettingDto userSettingDto) throws UserNotFoundException, GeneralException;
}
